### PR TITLE
[hab3_merge] Multi-Agent Trainers

### PR DESCRIPTION
**Pull request by ASzot**
## Motivation and Context

Merging the multi-agent trainers and setup.

## How Has This Been Tested

Hab3 evaluation matches `SIRo_dev`. Also tested end-to-end rearrangement policy training. 

## Types of changes

- **\[Docs change\]** Addition or changes to the documentation
- **\[Refactoring\]** Large changes to the code that improve its functionality or performance
- **\[Dependency Upgrade\]** Upgrades one or several dependencies in habitat
- **\[Bug Fix\]** (non-breaking change which fixes an issue)
- **\[Development\]** A pull request that add new features to the [habitat-lab](/habitat-lab) task and environment codebase. Development Pull Requests must be small (less that 500 lines of code change), have unit testing, very extensive documentation and examples. These are typically new tasks, environments, sensors, etc... The review process for these Pull Request is longer because these changes will be maintained by our core team of developers, so make sure your changes are easy to understand!
- **\[Experiment\]** A pull request that add new features to the [habitat-baselines](/habitat-baselines/) training codebase. Experiments Pull Requests can be any size, must have smoke/integration tests and be isolated from the rest of the code. Your code additions should not rely on other habitat-baselines code. This is to avoid dependencies between different parts of habitat-baselines. You must also include a README that will document how to use your new feature or trainer. **You** will be the maintainer of this code, if the code becomes stale or is not supported often enough, we will eventually remove it.

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated the documentation if required.
- [ ] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes if required.

### CODE REVIEW

**Comment by rpartsey** [comment]
Thanks for working on this PR @xavierpuigf and @ASzot!

I'm trying to make HITL tool work on top of this branch. I few things I've noticed  that are different compared to `SIRo_hitl`:
1. `HumanoidJointActionConfig` doesn't have `agent_index` argument.
2. `HumanoidRearrangeController` implementation is not up to date with `HumanoidRearrangeController` on `SIRo_hilt`. Shouldn't `HumanoidRearrangeController` on this branch be the same as on `SIRo_hitl`?
 - there is no `set_framerate_for_linspeed` 
 - `reset` is different
 - \+ some other code is different
 3. `SingleAgentAccessMgr`'s `resume_state`  argument doesn't default to None as in `MultiAgentAccessMgr`. Let's assign default value to None as well.
 4. `SingleAgentAccessMgr.policy_action_space_shape_lens` attribute is not defined.
 5. `hidden_state_shape_lens` is now attribute of the policy (actor_critic) not agent access manager.
 6.  There is not `oracle_nav_with_backing_up_action` action.
 7. HumanoidJointActionConfig.num_joints defaults to 17 (54 on SIRo_hitl).
 8. `OracleNavAction` implementation is not up to date with `OracleNavAction` on `SIRo_hilt`. For example, `lazy_inst_humanoid_controller` method or in `_get_target_for_idx` we should call `self.humanoid_controller.reset` with `self.cur_articulated_agent.base_transformation` not `self.cur_articulated_agent.base_pos`.
 9. Also, I don't understand why we don't pass `space_lengths` to `actor_critic.act`?
 10. I assume Spot's nav skill implementation is not fully ported since as you may see in [the video](https://cvmlp.slack.com/archives/C05HWV5UZ1A/p1695999069530579) it starts flying when it hits obstacles.
 
Please see the PR https://github.com/facebookresearch/habitat-lab/pull/1592 I created to highlight the changes needed to run HITL on top of `hab3_multi_agent`.

**Comment by rpartsey** [comment]
I've rebased my branch on top of `hab3_multi_agent` but I'm still seeing that not all the needed functionality was ported.

1. `HumanoidRearrangeController` implementation is not up to date with `HumanoidRearrangeController` on `SIRo_hilt`. Shouldn't `HumanoidRearrangeController` on this branch be the same as on `SIRo_hitl`?
2. `SingleAgentAccessMgr`'s `resume_state`  argument doesn't default to None as in `MultiAgentAccessMgr`. Let's assign default value to None as well.
3. `SingleAgentAccessMgr.policy_action_space_shape_lens` attribute is not defined.
4.  There is no `oracle_nav_with_backing_up_action` action.
5. HumanoidJointActionConfig.num_joints defaults to 17 (54 on SIRo_hitl).
6. `HumanoidJointAction` implementation is not up to date with `HumanoidJointAction` on `SIRo_hilt`. 
 
Please see the PR https://github.com/facebookresearch/habitat-lab/pull/1592 I created to highlight the changes needed to run HITL on top of `hab3_multi_agent` . 

Before `hab3_multi_agent` rebase I was able to launch HITL tool for all the app  states by applying changes from https://github.com/facebookresearch/habitat-lab/pull/1592 PR. But now I'm getting `AssertionError: ESP_CHECK failed: BulletArticulatedObject::updateJointMotor - No motor exists with motorId = 20`. See https://github.com/facebookresearch/habitat-lab/pull/1592 PR description for more details.

CC: @xavierpuigf

File: habitat-lab/habitat/tasks/rearrange/rearrange_sim.py [hunk 1]
```diff
@@ -640,9 +640,10 @@ def _add_objs(
                 ao = ao_mgr.get_object_by_handle(aoi_handle)
                 if self._kinematic_mode:
                     ao.motion_type = habitat_sim.physics.MotionType.KINEMATIC
-                    # remove any existing motors when converting to kinematic AO
-                    for motor_id in ao.existing_joint_motor_ids:
-                        ao.remove_joint_motor(motor_id)
+                    # TODO: figure out if we should uncomment that @aclegg
```
Comment (ASzot) [inline habitat-lab/habitat/tasks/rearrange/rearrange_sim.py:643]: I think Alex recently added this for better performance? Why comment it out?

Comment (xavierpuigf) [reply habitat-lab/habitat/tasks/rearrange/rearrange_sim.py:645]: The robot uses the joints to move the arm. If we dont comment out, we get the issue Ruslan was saying, where the joint 20 does not exist.

Comment (xavierpuigf) [reply habitat-lab/habitat/tasks/rearrange/rearrange_sim.py:645]: @rpartsey you can verify that this should not create the issue with joint motor 20

File: habitat-lab/habitat/tasks/rearrange/rearrange_task.py [hunk 2]
```diff
@@ -180,11 +184,38 @@ def _set_articulated_agent_start(self, agent_idx: int) -> None:
             agent_idx
         )
         if articulated_agent_start is None:
+            filter_func = None
```
Comment (akshararai) [inline habitat-lab/habitat/tasks/rearrange/rearrange_task.py:187]: ideally we should have better names for these variables

File: habitat-lab/habitat/tasks/rearrange/utils.py [hunk 3]
```diff
@@ -426,7 +426,121 @@ def write_gfx_replay(gfx_keyframe_str, task_config, ep_id):
...
+    agent: Optional[MobileManipulator] = None,
+    navmesh_offset: Optional[List[Tuple[float, float]]] = None,
+):
+    """
+    Places the robot at closest point if distance_threshold is -1.0 otherwise
+    will place the robot at `distance_threshold` away.
+    """
+    if distance_threshold == -1.0:
+        if navmesh_offset is not None:
+            return place_robot_at_closest_point_with_navmesh(
```
Comment (akshararai) [inline habitat-lab/habitat/tasks/rearrange/utils.py:445]: make sure this is using the new snapping function that Alex pushed?

**Review by akshararai — APPROVED** [review]
Mostly minor comments. A lot of these functions can benefit from docstrings, especially in multi_agent_access_mgr, pop_play_wrapper, etc. But we can add these in the next release.

File: habitat-baselines/habitat_baselines/config/habitat_baselines/rl/policy/hierarchical_policy/defined_skills/oracle_skills_human_multi_agent.yaml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+close_fridge:
+  skill_name: "NoopSkillPolicy"
+  obs_skill_inputs: ["obj_start_sensor"]
+  max_skill_steps: 1
+  apply_postconds: True
+  force_end_on_timeout: False
+  pddl_action_names: ["close_fridge_by_name"]
+  ignore_grip: True
+
+pick:
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/config/habitat_baselines/rl/policy/hierarchical_policy/defined_skills/oracle_skills_human_multi_agent.yaml:37]: `pddl_action_names` is only required when the skill name doesn't exactly correspond to a PDDL name in the PDDL domain.

File: habitat-baselines/habitat_baselines/config/habitat_baselines/rl/policy/hierarchical_policy/defined_skills/oracle_skills_multi_agent.yaml [hunk 5]
```diff
@@ -0,0 +1,93 @@
+# @package habitat_baselines.rl.policy.hierarchical_policy.defined_skills
+# See `oracle_skills.yaml` for details.
+
+open_cab:
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/config/habitat_baselines/rl/policy/hierarchical_policy/defined_skills/oracle_skills_multi_agent.yaml:4]: I personally prefer repeating but having the configs decoupled. The inheritance makes it more confusing I think.

File: habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+                active_agent_types,
+            )
+            self._multi_updater.set_active(
+                [self._agents[i].updater for i in self._active_agents]
+            )
+        self._multi_policy.set_active(
+            [self._agents[i].actor_critic for i in self._active_agents]
+        )
+
+    def post_init(self, create_rollouts_fn=None):
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py:303]: Docstring is in the parent class (this is a derived method). How should we indicate this?

File: habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        for agent in self._agents:
+            agent.load_ckpt_state_dict(ckpt)
+
+    @property
+    def masks_shape(self):
+        return (
+            sum(self._agents[i].masks_shape[0] for i in self._active_agents),
+        )
+
+    def after_update(self):
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py:340]: This is also inherited from the parent class. Same question as above.

File: habitat-lab/habitat/tasks/rearrange/multi_agent_sensors.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        # z coordinate is the height.
+        dist = ((rel_pos[0] ** 2) / (self._x_len**2)) + (
+            (rel_pos[1] ** 2) / (self._y_len**2)
+        )
+        return np.array([dist < 1], dtype=np.float32)
+
+
+@registry.register_measure
+class CooperateSubgoalReward(PddlSubgoalReward):
+    """
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-lab/habitat/tasks/rearrange/multi_agent_sensors.py:206]: This is actually the overall reward, not just the subgoal reward. I will rename it to `RearrangeCooperateReward`.

File: habitat-lab/habitat/tasks/rearrange/multi_task/domain_configs/fp.yaml [hunk 9]
```diff
@@ -0,0 +1,229 @@
+types:
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-lab/habitat/tasks/rearrange/multi_task/domain_configs/fp.yaml:1]: This isn't through Hydra so unfortunately it's not possible to compose.

File: habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+                )
+                cur_rollout.buffers[0] = prev_rollout.buffers[0]
+
+    def pre_rollout(self):
+        for agent in self._agents:
+            agent.pre_rollout()
+
+    def get_resume_state(self):
+        return {
+            agent_i: agent.get_resume_state()
```
Comment (jimmytyyang) [inline habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py:382]: the `agent_i` here should be string for dict?

**Review by jimmytyyang — COMMENTED** [review]
Hi Andrew, I catched a small bug in the PR. Please let me know what you think!

**Pull request merged**